versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,141 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.dremio.nessie.versioned.store.mongodb;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.dremio.nessie.versioned.ReferenceNotFoundException;
+import com.dremio.nessie.versioned.impl.InternalRef;
+import com.dremio.nessie.versioned.impl.condition.ConditionExpression;
+import com.dremio.nessie.versioned.impl.condition.UpdateExpression;
+import com.dremio.nessie.versioned.store.Id;
+import com.dremio.nessie.versioned.store.LoadOp;
+import com.dremio.nessie.versioned.store.LoadStep;
+import com.dremio.nessie.versioned.store.SaveOp;
+import com.dremio.nessie.versioned.store.Store;
+import com.dremio.nessie.versioned.store.ValueType;
+import com.google.common.collect.ListMultimap;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoDatabase;
+
+/**
+ * This class implements the Store interface that is used by Nessie as a backing store for versioning of it's
+ * Git like behaviour.
+ * The MongoDbStore connects to an external MongoDB server.
+ */
+public class MongoDbStore implements Store {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbStore.class);
+
+  private final String connectionString;
+  private final String databaseName;
+
+  // The client that connects to the MongoDB server.
+  protected MongoClient mongoClient;
+  // The database hosted by the MongoDB server.
+  private MongoDatabase mongoDatabase;
+
+  /**
+   * Creates a store ready for connection to a MongoDB instance.
+   * @param connectionString the URL of the database with which to connect. Eg mongodb://localhost:27017.
+   * @param databaseName the name of the database to retrieve
+   */
+  public MongoDbStore(String connectionString, String databaseName) {
+    this.connectionString = connectionString;
+    this.databaseName = databaseName;
+  }
+
+  /**
+   * Creates a connection to an existing database or creates the database if it does not exist.
+   * Since MongoDB creates databases and collections if they do not exist, there is no need to validate the presence of
+   * either before they are used.
+   */
+  @Override
+  public void start() {
+    mongoClient = MongoClients.create(connectionString);
+    mongoDatabase = mongoClient.getDatabase(databaseName);
+  }
+
+  /**
+   * Closes the connection this manager creates to a database. If the connection is already closed this method has
+   * no effect.
+   */
+  @Override
+  public void close() {
+    if (mongoClient != null) {
+      mongoClient.close();
+      mongoClient = null;
+    }
```

Comment:
MongoClient implements the Closeable and AutoCloseable interfaces, why not use those instead of calling close explicitly? Or why not use Try-with-resource, it should automatically close the MongoClient once it is out of scope.

Reply:
Is this not calling the close() method defined by AutoCloseable?

Reply:
I see what you mean. I now use try-with-resource.

Reply:
try-with-resource doesn't appear to make sense here - there is an explicit close and try-with-resource is for function or block scoped items. The reference to AutoCloseables.close() is an reference to a utility method which is not exposed here, given that I would say that this is fine.

Reply:
I did re-implement using try-with-resource. Given that I need the MongoClient for the methods that have not been implemented yet, I'm starting to see that try-with-resource makes it cumbersome as I keep having to instantiate the resource for each method. Sorry @keeratsingh I'll revert this change and go with @kylep-dremio suggestion.